test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java

```diff
@@ -349,4 +367,61 @@ private String formatStatusCodeMessage(ClientYamlTestResponse restTestResponse,
                 not(equalTo(408)),
                 not(equalTo(409)))));
     }
+
+    private static NodeSelector buildNodeSelector(XContentLocation location, String name, String value) {
+        switch (name) {
+        case "version":
+            Version[] range = SkipSection.parseVersionRange(value);
+            return new NodeSelector() {
+                @Override
+                public void select(Iterable<Node> nodes) {
+                    for (Iterator<Node> itr = nodes.iterator(); itr.hasNext();) {
+                        Node node = itr.next();
+                        if (node.getVersion() == null) {
+                            throw new IllegalStateException("expected [version] metadata to be set but got "
+                                    + node);
+                        }
+                        Version version = Version.fromString(node.getVersion());
+                        if (false == (version.onOrAfter(range[0]) && version.onOrBefore(range[1]))) {
+                            itr.remove();
+                        }
+                    }
+                }
+
+                @Override
+                public String toString() {
+                    return "version between [" + range[0] + "] and [" + range[1] + "]";
+                }
+            };
+        default:
+            throw new IllegalArgumentException("unknown node_selector [" + name + "]");
+        }
+    }
+
+    /**
+     * Selector that composes two selectors, running the "right" most selector
+     * first and then running the "left" selector on the results of the "right"
+     * selector.
+     */
+    private static class ComposeNodeSelector implements NodeSelector {
```

Comment:
when do we need this again? In case multiple selectors are configured in a yaml test?

Reply:
Right. It just feels bad to write it like you can only do one. It probably is overkill though. But I expect I'll make a followup to this PR to add support for node attributes and we probably want those in yaml too. Or maybe we don't? I'm actually not sure now.

Reply:
Having looked at some of the BWC tests lately, we're totally going to use this sooner rather than later.

Reply:
fine by me